verilog/tools/preprocessor/verilog_preprocessor.cc

```diff
@@ -49,53 +61,163 @@ static absl::Status StripComments(const SubcommandArgsRange& args,
+      !status.ok()) {
+    std::cerr << "ERROR: passed file can't be open\n.";
+    return status;
+  }
+  verilog::VerilogPreprocess::Config config;
+  config.filter_branches = 1;
+  // config.expand_macros=1;
+  verilog::VerilogPreprocess preprocessor(config);
+  verilog::VerilogLexer lexer(source_contents);
+  verible::TokenSequence lexed_sequence;
+  for (lexer.DoNextToken(); !lexer.GetLastToken().isEOF();
+       lexer.DoNextToken()) {
+    // For now we will store the syntax tree tokens only, ignoring all the
+    // white-space characters. however that should be stored to output the
+    // source code just like it was, but with conditionals filtered.
+    if (verilog::VerilogLexer::KeepSyntaxTreeTokens(lexer.GetLastToken()))
+      lexed_sequence.push_back(lexer.GetLastToken());
+  }
+  verible::TokenStreamView lexed_streamview;
+  // Initializing the lexed token stream view.
+  InitTokenStreamView(lexed_sequence, &lexed_streamview);
+  verilog::VerilogPreprocessData preprocessed_data =
+      preprocessor.ScanStream(lexed_streamview);
+  auto& preprocessed_stream = preprocessed_data.preprocessed_token_stream;
+  for (auto u : preprocessed_stream) outs << *u << '\n';
+  // for debugging.
+  for (auto& u : preprocessed_data.errors) outs << u.error_message << '\n';
+
+  //  parsing just as a trial
+  std::string post_preproc;
+  for (auto u : preprocessed_stream) post_preproc += std::string{u->text()};
+  std::string source_view{post_preproc};
+  verilog::VerilogAnalyzer analyzer(source_view, "file1", config);
+  auto analyze_status = analyzer.Analyze();
+  /* const auto& mydata = analyzer.Data().Contents(); */
+  /* outs<<mydata; */
 
-Input:
-  'file' is a Verilog or SystemVerilog source file.
-  Use '-' to read from stdin.
+  /* TODO(karimtera): regarding conditionals
+   1) Modify VerilogPreprocess config to have a configuration to generate SV
+   source codes for all possible variants. 2) Then use parser, directly from
+   VerilogAnalyzer or from VerilogParser to have less dependences. 3) Now, we
+   should have multiple trees, we need to merge them as described by Tom in
+   Verible's issue. 4) Finally, travese the tree and output the chosen path
+   based on definitions.
+  */
+  return absl::OkStatus();
+}
 
-Output: (stdout)
-  Contents of original file with // and /**/ comments removed.
-)"}},
-};
+static absl::Status GenerateVariants(const char* source_file, std::istream&,
+                                     std::ostream& outs, std::ostream&) {
+  std::string source_contents;
+  if (auto status = verible::file::GetContents(source_file, &source_contents);
+      !status.ok()) {
+    std::cerr << "ERROR: passed file can't be open\n.";
+    return status;  // check if the the file is not readable or doesn't exist.
+  }
 
-int main(int argc, char* argv[]) {
-  // Create a registry of subcommands (locally, rather than as a static global).
-  verible::SubcommandRegistry commands;
```

Comment:
You have removed the whole subcommand registry stuff. That should be kept.

Reply:
Yes sorry, but why would we need it?
I think using 'FileList' from 'verilog_project.h' for positional arguments, and `absl/flags` for other flags is enough.
I don't understand the difference between treating 'strip_comments' as a subcommand, or as a boolean flag.

Reply:
the subcommand is something where we can nicely implement features, such as listing them with their own description, have its own flags, can do its own tests (e.g. make sure that only one is there at a time), create separate manpages etc. (consider it similar to the idea of the git subcomments. like 'git push', 'git commit' etc are all subcommands.

You can get an inferior subset of the same functionality by using absl-flags, but since we already have Subcommands, why should we downgrade ?

Reply:
Ah okay, I get it.

I will have a deeper look into the subcommand class, and use it instead of the absl flags. 🙏🏻